internal/services/repair_manager/ruler/leaf.go

```
@@ -0,0 +1,184 @@
+	}
+
+	return &Leaf{
+		vtype:  ValueBool,
+		numVal: uint64(num),
+		strVal: "",
+	}
+}
+
+// NewLeafInt32 creates a new Leaf entry containing the supplied integer value.
+func NewLeafInt32(val int32) *Leaf {
+	return &Leaf{
+		vtype:  ValueInt32,
+		numVal: uint64(val),
+		strVal: "",
+	}
+}
+
+// NewLeafInt64 creates a new Leaf entry containing the supplied integer value.
+func NewLeafInt64(val int64) *Leaf {
+	return &Leaf{
+		vtype:  ValueInt64,
+		numVal: uint64(val),
+		strVal: "",
+	}
+}
+
+// --- Constructor functions
+
+// Evaluate returns this Leaf element.
+func (v *Leaf) Evaluate(*EvalContext) (*Leaf, error) { return v, nil }
+
+// Format returns a descriptive string for this instance.
+func (v *Leaf) Format(indent string) string {
+	return fmt.Sprintf("%s[Leaf: %s] %q", indent, v.typeName(), v.valueString())
+}
+
+// AsName returns a string where the replacement tokens are applied.  In order
+// to do so, this function accepts an array of strings that are alternating
+// strings of tokens and replacement values.
+func (v *Leaf) AsName(replacements []string) (string, error) {
+	if v.vtype != ValueString {
+		return v.AsString()
+	}
+
+	r := strings.NewReplacer(replacements...)
+
+	return r.Replace(v.strVal), nil
+}
+
+// AsInt64 returns the int64 value, if it is a numeric value.  int32 and int64
+// values are returned as expected, boolean returns 1 for true, 0 for false,
+// while string values return an error.
+func (v *Leaf) AsInt64() (int64, error) {
+	switch v.vtype {
+	case ValueBool, ValueInt64, ValueInt32:
+		return int64(v.numVal), nil
+
+	default:
+		return 0, ErrInvalidType
+	}
+}
+
+// AsInt32 returns the int632 value, if it is a numeric value.  int32 and int64
+// values are returned as expected (including truncation), boolean returns 1
+// for true, 0 for false, while string values return an error.
+func (v *Leaf) AsInt32() (int32, error) {
+	switch v.vtype {
+	case ValueBool, ValueInt64, ValueInt32:
+		return int32(v.numVal), nil
```

Hhmm. It's really ok to convert an int64 that's greater then ~2^32 to int32? It's the kind of thing that makes me nervous.

---

Yup, according the doc.  It gives you the lowest 32 bits and silently ignores any overflow.   Not sure that I like that, but it's what it says it does.